### PR TITLE
Commbase streamcache

### PR DESCRIPTION
**Pull request by dbollweg**
Opening draft PR for HIP improvements and commbase refactoring. I will go through the changes and guard the AMD changes so  to not change Nvidia performance.

### CODE REVIEW

**Comment by dbollweg** [comment]
Tests are passing now

**Review by lukas-mazur — CHANGES_REQUESTED** [review]
-

File: src/base/gutils.h [hunk 1]
```diff
@@ -11,7 +11,9 @@
 #include "wrapper/gpu_wrapper.h"
 #include <iostream>
 #include <math.h>
-
+#ifdef USE_NCCL
+#include <rccl.h>
```
[the remark this answers is not in the document]
Comment (dbollweg) [reply src/base/gutils.h:15]: It is not used at the moment.

[the remark this answers is not in the document]
Comment (lukas-mazur) [reply src/base/gutils.h:15]: Alright

**Review by lukas-mazur — APPROVED** [review]

**Pull request merged**